### PR TITLE
Add support to use predicate to stop event loop

### PR DESCRIPTION
**Pull request by mo-xiaoming**

### CODE REVIEW

File: include/hobbes/events/events.H [hunk 1]
```diff
@@ -29,6 +29,8 @@ void runEventLoop();
 // run the event loop for some number of microseconds
 void runEventLoop(int microsecondDuration);
```
Comment (smunix) [inline include/hobbes/events/events.H:30]: What do you think of the following instead?
```c++
void runEventLoop(int microsecondDuration, std::function<bool()> const& stopFn = []() -> bool { return false; })
```

File: include/hobbes/events/events.H [hunk 2]
```diff
@@ -29,6 +29,8 @@ void runEventLoop();
...
 
+// run the event loop until stopF() returns true
+void runEventLoop(std::function<bool()> &stopF);
```
Comment (smunix) [inline include/hobbes/events/events.H:33]: let's remove this

File: lib/hobbes/events/events.C [hunk 3]
```diff
@@ -91,6 +91,21 @@ void registerInterruptHandler(const std::function<void()>& fn) {
...
+      auto* c = reinterpret_cast<eventcbclosure*>(evts[fd].data.ptr);
+      (c->fn)(c->fd);
+      resetMemoryPool();
+    }
+  }
+  return fds;
+}
+}
+
 bool stepEventLoop(int timeoutMS) {
```
Comment (smunix) [inline lib/hobbes/events/events.C:109]: change the interface in a similar way,
```c++
bool stepEvenLoop(int timeoutMS, std::function<bool()> = []() -> bool { return false; })
```

File: lib/hobbes/events/events.C [hunk 4]
```diff
@@ -91,6 +91,21 @@ void registerInterruptHandler(const std::function<void()>& fn) {
   (*epClosures)[-1] = c;
 }
 
+namespace {
```
Comment (smunix) [inline lib/hobbes/events/events.C:94]: let's remove this

File: lib/hobbes/events/events.C [hunk 5]
```diff
@@ -91,6 +91,21 @@ void registerInterruptHandler(const std::function<void()>& fn) {
...
+      (c->fn)(c->fd);
+      resetMemoryPool();
+    }
+  }
+  return fds;
+}
+}
+
 bool stepEventLoop(int timeoutMS) {
   while (true) {
```
Comment (smunix) [inline lib/hobbes/events/events.C:110]:
```c++
while(not stopFn()) {
  ...
}
```

File: lib/hobbes/events/events.C [hunk 6]
```diff
@@ -170,19 +177,18 @@ void runEventLoop(int microsecondDuration) {
...
-        resetMemoryPool();
-      }
-    }
+    handleEvents(timeout);
+
     t = hobbes::time();
   } while (t < tf);
 }
 
+void runEventLoop(std::function<bool()> &stopF) {
```
Comment (smunix) [inline lib/hobbes/events/events.C:186]: revert this hunk

File: lib/hobbes/events/events.C [hunk 7]
```diff
@@ -100,16 +115,8 @@ bool stepEventLoop(int timeoutMS) {
...
-    int fds = epoll_wait(threadEPollFD(), evts, sizeof(evts)/sizeof(evts[0]), timeoutMS);
     bool status = true;
-    if (fds > 0) {
-      for (int fd = 0; fd < fds; ++fd) {
-        auto* c = reinterpret_cast<eventcbclosure*>(evts[fd].data.ptr);
-        (c->fn)(c->fd);
-        resetMemoryPool();
-      }
-    } else if (fds < 0) {
+    if (handleEvents(timeoutMS) < 0) {
```
Comment (smunix) [inline lib/hobbes/events/events.C:119]: revert this

File: lib/hobbes/events/events.C [hunk 8]
```diff
@@ -170,19 +177,18 @@ void runEventLoop(int microsecondDuration) {
...
-    struct epoll_event evts[64];
-    int fds = epoll_wait(threadEPollFD(), evts, sizeof(evts)/sizeof(evts[0]), timeout);
-    if (fds > 0) {
-      for (int fd = 0; fd < fds; ++fd) {
-        auto* c = reinterpret_cast<eventcbclosure*>(evts[fd].data.ptr);
-        (c->fn)(c->fd);
-        resetMemoryPool();
-      }
-    }
+    handleEvents(timeout);
```
Comment (smunix) [inline lib/hobbes/events/events.C:180]: revert

File: lib/hobbes/events/events.C (hunk 5 above)
Comment (mo-xiaoming) [reply lib/hobbes/events/events.C:110]: `epoll_wait` blocks forever on default `timeoutMS` value `-1`, `stopFn` won't get the second chance to be evaluated

Comment (smunix) [reply lib/hobbes/events/events.C:110]: The timeout delay used by `epoll_wait` can be changed by pushing a timer event; there's an api for that

**Review by smunix — APPROVED** [review]
nice! thank you @mo-xiaoming

**Pull request merged**